src/meta/server_state.cpp

```diff
@@ -2943,6 +2944,117 @@ bool server_state::validate_target_max_replica_count(int32_t max_replica_count)
+
+    std::map<std::string, std::string> envs;
+    std::string app_path = "";
+    app_info ainfo;
+    {
+        zauto_read_lock l(_lock);
+        std::shared_ptr<app_state> app = get_app(app_name);
+        if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
+            response.err = app == nullptr ? ERR_APP_NOT_EXIST : ERR_APP_DROPPED;
+            response.hint_msg =
+                fmt::format("app {} is {}",
+                            app_name,
+                            response.err == ERR_APP_NOT_EXIST ? "not existed" : "not available");
+            derror_f("{}", response.hint_msg);
+            return;
+        }
+        envs = app->envs;
+        ainfo = *(reinterpret_cast<app_info *>(app.get()));
+        app_path = get_app_path(*app);
+    }
+
+    auto iter = envs.find(replica_envs::MANUAL_COMPACT_DISABLED);
+    if (iter != envs.end() && iter->second == "true") {
+        response.err = ERR_OPERATION_DISABLED;
+        response.hint_msg = fmt::format("app {} disable manual compaction", app_name);
+        derror_f("{}", response.hint_msg);
+        return;
+    }
+
+    // parse manual compaction envs
+    std::vector<std::string> keys;
+    std::vector<std::string> values;
+
+    int32_t target_level = -1;
+    if (request.__isset.target_level) {
+        target_level = request.target_level;
+        if (target_level < -1) {
+            response.err = ERR_INVALID_PARAMETERS;
+            response.hint_msg = fmt::format(
+                "invalid target_level({}), should in range of [-1, num_levels]", target_level);
+            derror_f("{}", response.hint_msg);
+            return;
+        }
+    }
+    keys.emplace_back(replica_envs::MANUAL_COMPACT_ONCE_TARGET_LEVEL);
+    values.emplace_back(std::to_string(target_level));
+
+    if (request.__isset.max_running_count) {
+        if (request.max_running_count < 0) {
+            response.err = ERR_INVALID_PARAMETERS;
+            response.hint_msg =
+                fmt::format("invalid max_running_count({}), should be greater than 0",
+                            request.max_running_count);
+            derror_f("{}", response.hint_msg);
+            return;
+        }
+        if (request.max_running_count > 0) {
+            keys.emplace_back(replica_envs::MANUAL_COMPACT_MAX_CONCURRENT_RUNNING_COUNT);
+            values.emplace_back(std::to_string(request.max_running_count));
+        }
+    }
+
+    std::string bottommost = "skip";
+    if (request.__isset.bottommost && request.bottommost) {
+        bottommost = "force";
+    }
+    keys.emplace_back(replica_envs::MANUAL_COMPACT_ONCE_BOTTOMMOST_LEVEL_COMPACTION);
+    values.emplace_back(bottommost);
+
+    int64_t trigger_time = dsn_now_s();
+    if (request.__isset.trigger_time) {
+        trigger_time = request.trigger_time;
+    }
+    keys.emplace_back(replica_envs::MANUAL_COMPACT_ONCE_TRIGGER_TIME);
+    values.emplace_back(std::to_string(trigger_time));
+
+    // update new app_envs
```

Comment:
Good annotation is used for explaining why did you do like this, not what to do here.

Reply:
I sometimes agree with you. In this case, the annotation is used as a division. But I also update it in my new commit.